### PR TITLE
riscv/virt: support graceful shutdown via SBI

### PR DESCRIPTION
**Pull request by yf13**
## Summary

This patch enables graceful shutdown command `poweroff` in kernel mode
configs `rv-virt/knsh64` and `rv-virt/knsh32`. The implementation is
based on SBI v0.3 compliant firmware (supported in opensbi v1.0).

## Impact

risc-v kernel builds

## Testing

- checked `rv-virt/knsh32`  and `rv-virt/knsh64` with qemu v6.2 and opensbi v1.0 on Ubuntu 22.04
- CI checks

### CODE REVIEW

File: arch/risc-v/src/common/supervisor/riscv_sbi.c [hunk 1]
```diff
@@ -132,6 +132,12 @@ uintptr_t riscv_sbi_send_ipi(uint32_t hmask, uintptr_t hbase)
 }
 
 #ifndef CONFIG_NUTTSBI
+uintptr_t riscv_sbi_system_reset(uint32_t type, uint32_t reason)
+{
+  return sbi_ecall(SBI_EXT_SRST, SBI_EXT_SRST_SYS_RESET,
+      type, reason, 0, 0, 0, 0);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/supervisor/riscv_sbi.c:138]: align

File: boards/risc-v/qemu-rv/rv-virt/src/qemu_rv_appinit.c [hunk 2]
```diff
@@ -35,6 +35,7 @@
 #include <nuttx/virtio/virtio-mmio.h>
 
 #include <sys/mount.h>
+#include "riscv_internal.h"
```
Comment (xiaoxiang781216) [inline boards/risc-v/qemu-rv/rv-virt/src/qemu_rv_appinit.c:39]: move blank before `#include "riscv_internal.h"`

File: arch/risc-v/src/common/riscv_internal.h [hunk 3]
```diff
@@ -335,6 +351,7 @@ void riscv_sbi_set_timer(uint64_t stime_value);
 uint64_t riscv_sbi_get_time(void);
 uintptr_t riscv_sbi_boot_secondary(uint32_t hartid, uintptr_t addr,
                                    uintptr_t a1);
+uintptr_t riscv_sbi_system_reset(uint32_t type, uint32_t reason);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_internal.h:354]: what about non-S mode?

Comment (yf13) [reply arch/risc-v/src/common/riscv_internal.h:354]: There seems no standard way to do that in non-S mode. For `rv-virt` target, more investigation is needed to see how to do it in M-mode later.

File: boards/risc-v/qemu-rv/rv-virt/src/qemu_rv_appinit.c (hunk 2 above)
Comment (yf13) [reply boards/risc-v/qemu-rv/rv-virt/src/qemu_rv_appinit.c:38]: blank line added before it now,

File: arch/risc-v/src/common/riscv_internal.h (hunk 3 above)
Comment (xiaoxiang781216) [reply arch/risc-v/src/common/riscv_internal.h:354]: OK

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

**Review by anchao — APPROVED** [review]

**Pull request merged**